src/bindgen/cargo/cargo_expand.rs

```diff
@@ -37,39 +34,31 @@ impl From<Utf8Error> for Error {
 /// removing any macros in the process.
 pub fn expand(manifest_path: &Path, crate_name: &str, version: &str) -> Result<String, Error> {
     let cargo = env::var("CARGO").unwrap_or_else(|_| String::from("cargo"));
-    let run = |target_dir: &Path| {
-        let mut cmd = Command::new(cargo);
-        cmd.env("CARGO_TARGET_DIR", target_dir);
-        cmd.arg("rustc");
-        cmd.arg("--lib");
-        cmd.arg("--manifest-path");
-        cmd.arg(manifest_path);
-        cmd.arg("--all-features");
-        cmd.arg("-p");
-        cmd.arg(&format!("{}:{}", crate_name, version));
-        cmd.arg("--");
-        cmd.arg("-Z");
-        cmd.arg("unstable-options");
-        cmd.arg("--pretty=expanded");
-        let output = cmd.output()?;
+    let mut cmd = Command::new(cargo);
+
+    if let Ok(ref path) = env::var("CARGO_EXPAND_TARGET_DIR") {
+        cmd.env("CARGO_TARGET_DIR", path);
+    }
 
-        let src = from_utf8(&output.stdout)?.to_owned();
-        let error = from_utf8(&output.stderr)?.to_owned();
+    cmd.arg("rustc");
+    cmd.arg("--lib");
+    cmd.arg("--manifest-path");
+    cmd.arg(manifest_path);
+    cmd.arg("--all-features");
+    cmd.arg("-p");
+    cmd.arg(&format!("{}:{}", crate_name, version));
+    cmd.arg("--");
+    cmd.arg("-Z");
+    cmd.arg("unstable-options");
+    cmd.arg("--pretty=expanded");
+    let output = cmd.output()?;
 
-        if src.len() == 0 {
-            Err(Error::Compile(error))
-        } else {
-            Ok(src)
-        }
-    };
+    let src = from_utf8(&output.stdout)?.to_owned();
+    let error = from_utf8(&output.stderr)?.to_owned();
 
-    if let Ok(ref path) = env::var("CARGO_EXPAND_TARGET_DIR") {
-        run(&Path::new(path))
+    if src.len() == 0 {
```

Comment:
This can be true if `cargo rustc` is run twice consecutively. In that case `stderr` will be just "Finished in 0.0 seconds" and the exit code will be 0.

I'll add my fix in a separate commit, but there are several discussion-worthy solutions.

Reply:
@eqrion where is the best place to add tests for this?

Reply:
Actually I can't reproduce this problem at all on a different machine...